File: src/csharp/pulse/cdm/properties/SEScalarPressurePerVolume.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class PressurePerVolumeUnit : Unit
  {
    private PressurePerVolumeUnit(string v) : base(v) { }

    public static PressurePerVolumeUnit mmHg_Per_mL = new PressurePerVolumeUnit("mmHg/mL");
    public static PressurePerVolumeUnit cmH2O_Per_mL = new PressurePerVolumeUnit("cmH2O/mL");
    public static PressurePerVolumeUnit cmH2O_Per_L = new PressurePerVolumeUnit("cmH2O/L");
    
    public static PressurePerVolumeUnit FromString(string u)
    {
      if (u == PressurePerVolumeUnit.mmHg_Per_mL.ToString())
        return PressurePerVolumeUnit.mmHg_Per_mL;
      if (u == PressurePerVolumeUnit.cmH2O_Per_mL.ToString())
        return PressurePerVolumeUnit.cmH2O_Per_mL;
      if (u == PressurePerVolumeUnit.cmH2O_Per_L.ToString())
        return PressurePerVolumeUnit.cmH2O_Per_L;

      throw new System.ArgumentException(u + " is not a know pressure per volume unit");
    }
  }

  public class SEScalarPressurePerVolume : SEScalar
  {
    protected PressurePerVolumeUnit unit;

    public SEScalarPressurePerVolume() : base()
    {

    }

    public SEScalarPressurePerVolume(SEScalarPressurePerVolume from)
    {
      Set(from);
    }

    public SEScalarPressurePerVolume(double value, PressurePerVolumeUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarPressurePerVolume from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }

    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, PressurePerVolumeUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }

    public double GetValue(PressurePerVolumeUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public PressurePerVolumeUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}

File: src/cpp/cdm/properties/SEScalarPressurePerVolume.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarPressurePerVolume.h"

const PressurePerVolumeUnit PressurePerVolumeUnit::mmHg_Per_mL("mmHg/mL");
const PressurePerVolumeUnit PressurePerVolumeUnit::cmH2O_Per_mL("cmH2O/mL");
const PressurePerVolumeUnit PressurePerVolumeUnit::cmH2O_Per_L("cmH2O/L");

bool PressurePerVolumeUnit::IsValidUnit(const std::string& unit)
{
  if (mmHg_Per_mL.GetString().compare(unit) == 0)
    return true;
  if (cmH2O_Per_mL.GetString().compare(unit) == 0)
    return true;
  if (cmH2O_Per_L.GetString().compare(unit) == 0)
    return true;
  return false;
}

const PressurePerVolumeUnit& PressurePerVolumeUnit::GetCompoundUnit(const std::string& unit)
{
  if (mmHg_Per_mL.GetString().compare(unit) == 0)
    return mmHg_Per_mL;
  if (cmH2O_Per_mL.GetString().compare(unit) == 0)
    return cmH2O_Per_mL;
  if (cmH2O_Per_L.GetString().compare(unit) == 0)
    return cmH2O_Per_L;
  throw CommonDataModelException(unit + " is not a valid PressurePerVolume unit");
}

template class SEScalarQuantity<PressurePerVolumeUnit>;
